src/field/field.rs

```diff
@@ -486,3 +486,212 @@ macro_rules! test_square_root {
+        -> Result<()>
+    where
+        T: Field,
+        UnaryOp: Fn(T) -> T,
+        ExpectedOp: Fn(BigUint) -> BigUint
+    {
+        let inputs = test_inputs(modulus, word_bits);
+        let expected = inputs.iter().map(|x| expected_op(x.clone()));
+        let output = inputs.iter().map(|x| field_to_biguint(op(biguint_to_field(x.clone()))));
+        // Compare expected outputs with actual outputs
+        assert!(output.zip(expected).all(|(x, y)| x == y),
+                "output differs from expected");
+        Ok(())
+    }
+
+
+    /// Apply the binary functions `op` and `expected_op` to each pair
+    /// in `zip(inputs, rotate_right(inputs, i))` where `inputs` is
+    /// `test_inputs(modulus, word_bits)` and `i` ranges from 0 to
+    /// `inputs.len()`.  Panic if the two functions ever give
+    /// different answers.
+    pub fn run_binaryop_test_cases<T, BinaryOp, ExpectedOp>(
+        modulus: &BigUint, word_bits: usize,
+        op: BinaryOp, expected_op: ExpectedOp)
+        -> Result<()>
+    where
+        T: Field,
+        BinaryOp: Fn(T, T) -> T,
+        ExpectedOp: Fn(BigUint, BigUint) -> BigUint
+    {
+        let inputs = test_inputs(modulus, word_bits);
+
+        for i in 0 .. inputs.len() {
+            // Iterator over inputs rotated right by i places. Since
+            // cycle().skip(i) rotates left by i, we need to rotate by
+            // n_input_elts - i.
+            let shifted_inputs = inputs.iter().cycle().skip(inputs.len() - i);
+            // Calculate pointwise operations
+            let expected = inputs.iter().zip(shifted_inputs.clone()).map(
+                |(x, y)| expected_op(x.clone(), y.clone()));
+            let output = inputs.iter().zip(shifted_inputs).map(
+                |(x, y)| field_to_biguint(op(biguint_to_field(x.clone()),
+                                             biguint_to_field(y.clone()))));
+            // Compare expected outputs with actual outputs
+            assert!(output.zip(expected).all(|(x, y)| x == y),
+                    "output differs from expected at rotation {}", i);
+        }
+        Ok(())
+    }
+}
+
+#[macro_export]
+macro_rules! test_arithmetic {
+    ($field:ty) => {
+        use crate::field_tests;
+        use crate::field_to_biguint;
+        use std::io::Result;
+        use std::ops::{Add,Sub,Mul,Neg,Div};
+        use num::{BigUint, Zero, One};
+
+
+        /// Return the modulus of the type `Fld`.
+        fn field_modulus<Fld>() -> BigUint
+        where Fld: Field {
+            field_to_biguint(Fld::T) * (BigUint::one() << Fld::TWO_ADICITY) + 1u32
+        }
+
+        // Can be 32 or 64; doesn't have to be computer's actual word
+        // bits. Choosing 32 gives more tests...
+        const WORD_BITS: usize = 32;
+
+        // TODO: I've given these unreasonably long names to avoid
+        // name conflicts (notably with the 'negation()' test for
+        // Bls12377Base. It would be better that these tests were
```

Comment:
That's a good idea. I think it should work fine like

```
macro_rules! test_arithmetic {
    ($field:ty) => {
        mod arithmetic {
            ...
```

and besides method name collisions, it might help avoid redundant imports (not sure if those would cause compiler warnings or what).

Reply:
I wrapped the test functions in `mod arithmetic { ... }` and tried fixing up the various `use` statements and references to `$field` but I couldn't get this to work. Any clues as to what might be tripping me up?

Reply:
Got it working here :) #79 

I wasn't quite sure what to do with `test_square_root()`; for now I just wrote an import inside the method. Would it make sense to merge `test_square_root!` into `test_arithmetic!`?

Reply:
Yes, I think that makes sense. Feel free to do that in #79, otherwise I'm happy to do it here.